File: chapter5/chapt5_ws/src/demo_python_tf/demo_python_tf/tf_listener.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformListener, Buffer  # 坐标监听器、缓存器
from tf_transformations import euler_from_quaternion  # 四元数转欧拉角
import math  # 角度转弧度

class TFListener(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self._buffer = Buffer()
        self._broadcaster = TransformListener(self._buffer, self)  # self 就是节点本身
        self._timer =  self.create_timer(1.0, self.get_tf)  # 1秒 一秒获取1次tf

    def get_tf(self):  # 实时查询坐标关系
        try:
            # lookup_transform("base_link", "bottle_link", 查询哪个时刻的【0代表最新】, 超过1秒查不到就不查了)
            result = self._buffer.lookup_transform("base_link", "bottle_link", rclpy.time.Time(seconds=0.0), rclpy.time.Duration(seconds=1.0))
            transform = result.transform
            self.get_logger().info(f"平移：{transform.translation}")
            self.get_logger().info(f"旋转：{transform.rotation}")
            rotation_euler = euler_from_quaternion([
                transform.rotation.x,
                transform.rotation.y,
                transform.rotation.z,
                transform.rotation.w,
            ])
            self.get_logger().info(f"旋转[欧拉角]：{rotation_euler}")
        except Exception as e:
            self.get_logger().warn(f"获取坐标变换失败，原因：{e}")
        
def main():
    rclpy.init()
    node = TFListener("tf_listener")
    rclpy.spin(node)
    rclpy.shutdown()

File: chapter5/chapt5_ws/src/demo_cpp_tf/src/tf_listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"  // 提供消息接口
#include "tf2/LinearMath/Quaternion.hpp"  // 提供 tf2::Quaternion 类
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"  // 消息类型转换函数
#include "tf2_ros/transform_listener.hpp"  // 坐标监听器
#include "tf2_ros/buffer.hpp"
#include "tf2/utils.hpp"  // 四元数转欧拉角
#include "chrono"

using namespace std::chrono_literals;  // s ms

/*
ros2 topic list

ros2 topic echo /tf

ros2 topic info /tf -v

ros2 run tf2_ros tf2_echo base_link target_point
*/

class TFListener: public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::TransformListener> _listener;
    rclcpp::TimerBase::SharedPtr _timer;  // 定时器
    std::shared_ptr<tf2_ros::Buffer> _buffer;  // 将监听到的坐标关系存储下来

public:
    TFListener(std::string node_name): Node(node_name)
    {
        _buffer = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        _listener = std::make_shared<tf2_ros::TransformListener>(*_buffer, this);
        // 把成员函数当做回调函数，使用函数包装器 std::bind(&类::成员函数, 对象指针)
        _timer = this->create_wall_timer(1s, std::bind(&TFListener::getTransform, this));
    }

    void getTransform()
    {
        // 到 _buffer 里查询坐标关系
        try
        {
            // 查询从 target_point 到 base_link 的坐标关系
            const auto transform = _buffer->lookupTransform("base_link", "target_point", this->get_clock()->now(), rclcpp::Duration::from_seconds(1.0f));
        
            // 获取查询结果
            auto _translation = transform.transform.translation;
            auto _rotation = transform.transform.rotation;

            //  四元数转欧拉角
            double y, p, r;
            tf2::getEulerYPR(_rotation, y, p, r);
            RCLCPP_INFO(this->get_logger(), "平移：%f, %f, %f", _translation.x, _translation.y, _translation.z);
            RCLCPP_INFO(this->get_logger(), "旋转[欧拉角]：%f, %f, %f", y, p, r);

        }
        catch(const std::exception& e)
        {
            // 没有查询到坐标关系
            RCLCPP_WARN(this->get_logger(), "%s", e.what());
        }
        
    }

};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFListener>("tf_listener");
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}